File: FlowCrypt/src/main/java/com/flowcrypt/email/api/email/model/OutgoingMessageInfo.java
/*
 * © 2016-2018 FlowCrypt Limited. Limitations apply. Contact human@flowcrypt.com
 * Contributors: DenBond7
 */

package com.flowcrypt.email.api.email.model;

import android.os.Parcel;

import com.flowcrypt.email.js.PgpContact;
import com.flowcrypt.email.model.MessageEncryptionType;

import java.util.ArrayList;

/**
 * Simple POJO class which describe an outgoing message model.
 *
 * @author DenBond7
 * Date: 09.05.2017
 * Time: 11:20
 * E-mail: DenBond7@gmail.com
 */

public class OutgoingMessageInfo extends MessageInfo {

  public static final Creator<OutgoingMessageInfo> CREATOR = new Creator<OutgoingMessageInfo>() {
    @Override
    public OutgoingMessageInfo createFromParcel(Parcel source) {
      return new OutgoingMessageInfo(source);
    }

    @Override
    public OutgoingMessageInfo[] newArray(int size) {
      return new OutgoingMessageInfo[size];
    }
  };
  private PgpContact[] toPgpContacts;
  private PgpContact[] ccPgpContacts;
  private PgpContact[] bccPgpContacts;
  private PgpContact fromPgpContact;
  private String rawReplyMsg;
  private ArrayList<AttachmentInfo> atts;
  private ArrayList<AttachmentInfo> fwdAtts;
  private MessageEncryptionType encryptionType;
  private boolean isForwarded;
  private long uid;

  public OutgoingMessageInfo() {
  }

  protected OutgoingMessageInfo(Parcel in) {
    super(in);
    this.toPgpContacts = in.createTypedArray(PgpContact.CREATOR);
    this.ccPgpContacts = in.createTypedArray(PgpContact.CREATOR);
    this.bccPgpContacts = in.createTypedArray(PgpContact.CREATOR);
    this.fromPgpContact = in.readParcelable(PgpContact.class.getClassLoader());
    this.rawReplyMsg = in.readString();
    this.atts = in.createTypedArrayList(AttachmentInfo.CREATOR);
    this.fwdAtts = in.createTypedArrayList(AttachmentInfo.CREATOR);
    int tmpEncryptionType = in.readInt();
    this.encryptionType = tmpEncryptionType == -1 ? null : MessageEncryptionType.values()[tmpEncryptionType];
    this.isForwarded = in.readByte() != 0;
    this.uid = in.readLong();
  }

  @Override
  public int describeContents() {
    return 0;
  }

  @Override
  public void writeToParcel(Parcel dest, int flags) {
    super.writeToParcel(dest, flags);
    dest.writeTypedArray(this.toPgpContacts, flags);
    dest.writeTypedArray(this.ccPgpContacts, flags);
    dest.writeTypedArray(this.bccPgpContacts, flags);
    dest.writeParcelable(this.fromPgpContact, flags);
    dest.writeString(this.rawReplyMsg);
    dest.writeTypedList(this.atts);
    dest.writeTypedList(this.fwdAtts);
    dest.writeInt(this.encryptionType == null ? -1 : this.encryptionType.ordinal());
    dest.writeByte(this.isForwarded ? (byte) 1 : (byte) 0);
    dest.writeLong(this.uid);
  }

  public PgpContact[] getCcPgpContacts() {
    return ccPgpContacts;
  }

  public void setCcPgpContacts(PgpContact[] ccPgpContacts) {
    this.ccPgpContacts = ccPgpContacts;
  }

  public PgpContact[] getBccPgpContacts() {
    return bccPgpContacts;
  }

  public void setBccPgpContacts(PgpContact[] bccPgpContacts) {
    this.bccPgpContacts = bccPgpContacts;
  }

  public PgpContact[] getToPgpContacts() {
    return toPgpContacts;
  }

  public void setToPgpContacts(PgpContact[] toPgpContacts) {
    this.toPgpContacts = toPgpContacts;
  }

  public PgpContact getFromPgpContact() {
    return fromPgpContact;
  }

  public void setFromPgpContact(PgpContact fromPgpContact) {
    this.fromPgpContact = fromPgpContact;
  }

  public String getRawReplyMsg() {
    return rawReplyMsg;
  }

  public void setRawReplyMsg(String rawReplyMsg) {
    this.rawReplyMsg = rawReplyMsg;
  }

  public ArrayList<AttachmentInfo> getAttachments() {
    return atts;
  }

  public void setAttachments(ArrayList<AttachmentInfo> atts) {
    this.atts = atts;
  }

  public MessageEncryptionType getEncryptionType() {
    return encryptionType;
  }

  public void setEncryptionType(MessageEncryptionType encryptionType) {
    this.encryptionType = encryptionType;
  }

  public boolean isForwarded() {
    return isForwarded;
  }

  public void setForwarded(boolean forwarded) {
    isForwarded = forwarded;
  }

  public ArrayList<AttachmentInfo> getForwardedAttachments() {
    return fwdAtts;
  }

  public void setForwardedAttachments(ArrayList<AttachmentInfo> fwdAtts) {
    this.fwdAtts = fwdAtts;
  }

  public long getUid() {
    return uid;
  }

  public void setUid(long uid) {
    this.uid = uid;
  }
}

File: FlowCrypt/src/main/java/com/flowcrypt/email/api/email/model/OutgoingMessageInfo.kt
/*
 * © 2016-present FlowCrypt a.s. Limitations apply. Contact human@flowcrypt.com
 * Contributors: denbond7
 */

package com.flowcrypt.email.api.email.model

import android.os.Parcelable
import com.flowcrypt.email.database.MessageState
import com.flowcrypt.email.database.entity.MessageEntity
import com.flowcrypt.email.model.MessageEncryptionType
import com.flowcrypt.email.model.MessageType
import com.google.gson.annotations.Expose
import jakarta.mail.Flags
import jakarta.mail.internet.InternetAddress
import kotlinx.parcelize.IgnoredOnParcel
import kotlinx.parcelize.Parcelize
import java.util.UUID

/**
 * Simple POJO class which describe an outgoing message model.
 *
 * @author Denys Bondarenko
 */
@Parcelize
data class OutgoingMessageInfo(
  @Expose val account: String? = null,
  @Expose val accountType: String? = null,
  @Expose val subject: String? = null,
  @Expose val msg: String? = null,
  @Expose val toRecipients: List<InternetAddress>? = null,
  @Expose val ccRecipients: List<InternetAddress>? = null,
  @Expose val bccRecipients: List<InternetAddress>? = null,
  @Expose val from: InternetAddress? = null,
  @Expose val atts: List<AttachmentInfo>? = null,
  @Expose val forwardedAtts: List<AttachmentInfo>? = null,
  @Expose val encryptionType: MessageEncryptionType? = null,
  @Expose @MessageType val messageType: Int = MessageType.NEW,
  @Expose val replyToMessageEntityId: Long? = null,
  @Expose val uid: Long = 0,
  @Expose val password: CharArray? = null,
  @Expose val timestamp: Long = System.currentTimeMillis(),
  @Expose val signature: String? = null,
  @Expose val quotedTextForReply: String? = null,
  @Expose val draftId: String? = null,
) : Parcelable {

  @IgnoredOnParcel
  val isPasswordProtected = password?.isNotEmpty()

  /**
   * Generate a list of the all recipients.
   *
   * @return A list of the all recipients
   */
  fun getAllRecipients(): List<String> {
    return getPublicRecipients() + getProtectedRecipients()
  }

  fun getPublicRecipients(): List<String> {
    val toAddresses = toRecipients?.map { it.address } ?: emptyList()
    val ccAddresses = ccRecipients?.map { it.address } ?: emptyList()
    return toAddresses + ccAddresses
  }

  fun getProtectedRecipients() = bccRecipients?.map { it.address } ?: emptyList()
  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as OutgoingMessageInfo

    if (account != other.account) return false
    if (subject != other.subject) return false
    if (msg != other.msg) return false
    if (toRecipients != other.toRecipients) return false
    if (ccRecipients != other.ccRecipients) return false
    if (bccRecipients != other.bccRecipients) return false
    if (from != other.from) return false
    if (atts != other.atts) return false
    if (forwardedAtts != other.forwardedAtts) return false
    if (encryptionType != other.encryptionType) return false
    if (messageType != other.messageType) return false
    if (replyToMessageEntityId != other.replyToMessageEntityId) return false
    if (uid != other.uid) return false
    if (password != null) {
      if (other.password == null) return false
      if (!password.contentEquals(other.password)) return false
    } else if (other.password != null) return false
    if (timestamp != other.timestamp) return false
    if (signature != other.signature) return false
    if (quotedTextForReply != other.quotedTextForReply) return false
    if (draftId != other.draftId) return false
    return true
  }

  override fun hashCode(): Int {
    var result = account?.hashCode() ?: 0
    result = 31 * result + (subject?.hashCode() ?: 0)
    result = 31 * result + (msg?.hashCode() ?: 0)
    result = 31 * result + (toRecipients?.hashCode() ?: 0)
    result = 31 * result + (ccRecipients?.hashCode() ?: 0)
    result = 31 * result + (bccRecipients?.hashCode() ?: 0)
    result = 31 * result + (from?.hashCode() ?: 0)
    result = 31 * result + (atts?.hashCode() ?: 0)
    result = 31 * result + (forwardedAtts?.hashCode() ?: 0)
    result = 31 * result + (encryptionType?.hashCode() ?: 0)
    result = 31 * result + messageType
    result = 31 * result + (replyToMessageEntityId?.hashCode() ?: 0)
    result = 31 * result + uid.hashCode()
    result = 31 * result + (password?.contentHashCode() ?: 0)
    result = 31 * result + timestamp.hashCode()
    result = 31 * result + (signature?.hashCode() ?: 0)
    result = 31 * result + (quotedTextForReply?.hashCode() ?: 0)
    result = 31 * result + (draftId?.hashCode() ?: 0)
    return result
  }

  fun toMessageEntity(
    folder: String,
    flags: Flags,
    replyTo: String? = null,
    password: ByteArray? = null,
  ): MessageEntity {
    val timestamp = System.currentTimeMillis()
    return MessageEntity(
      account = requireNotNull(account),
      accountType = requireNotNull(accountType),
      folder = folder,
      uid = uid,
      receivedDate = timestamp,
      sentDate = timestamp,
      fromAddresses = from.toString(),
      replyToAddresses = replyTo,
      toAddresses = InternetAddress.toString(toRecipients?.toTypedArray()),
      ccAddresses = InternetAddress.toString(ccRecipients?.toTypedArray()),
      subject = subject,
      flags = flags.toString().uppercase(),
      hasAttachments = atts?.isNotEmpty() == true || forwardedAtts?.isNotEmpty() == true,
      isNew = false,
      isEncrypted = encryptionType == MessageEncryptionType.ENCRYPTED,
      state = if (messageType == MessageType.FORWARD) {
        MessageState.NEW_FORWARDED.value
      } else {
        MessageState.NEW.value
      },
      password = password,
      attachmentsDirectory = UUID.randomUUID().toString(),
      draftId = draftId
    )
  }
}
